crypto/cmp/cmp_genm.c

```diff
@@ -0,0 +1,183 @@
+                       "certificate from '%s' with subject '%s' %s",
+                       source, subj, msg);
+    OPENSSL_free(subj);
+}
+
+static int ossl_X509_check(OSSL_CMP_CTX *ctx, const char *source, X509 *cert,
+                           int type_CA, const X509_VERIFY_PARAM *vpm)
+{
+    uint32_t ex_flags = X509_get_extension_flags(cert);
+    int res = X509_cmp_timeframe(vpm, X509_get0_notBefore(cert),
+                                 X509_get0_notAfter(cert));
+    int ret = res == 0;
+    OSSL_CMP_severity level =
+        vpm == NULL ? OSSL_CMP_LOG_WARNING : OSSL_CMP_LOG_ERR;
+
+    if (!ret)
+        cert_msg(OPENSSL_FUNC, OPENSSL_FILE, OPENSSL_LINE, level, ctx,
+                 source, cert, res > 0 ? "has expired" : "not yet valid");
+    if (type_CA >= 0 && (ex_flags & EXFLAG_V1) == 0) {
+        int is_CA = (ex_flags & EXFLAG_CA) != 0;
+
+        if ((type_CA == 1) != is_CA) {
+            cert_msg(OPENSSL_FUNC, OPENSSL_FILE, OPENSSL_LINE, level, ctx,
+                     source, cert,
+                     is_CA ? "is not an EE cert" : "is not a CA cert");
+            ret = 0;
+        }
+    }
+    return ret;
+}
+
+static int ossl_X509_check_all(OSSL_CMP_CTX *ctx, const char *source,
+                               STACK_OF(X509) *certs,
+                               int type_CA, const X509_VERIFY_PARAM *vpm)
+{
+    int i;
+    int ret = 1;
+
+    for (i = 0; i < sk_X509_num(certs /* may be NULL */); i++)
+        ret = ossl_X509_check(ctx, source,
+                              sk_X509_value(certs, i), type_CA, vpm)
+            && ret; /* Having 'ret' after the '&&', all certs are checked. */
+    return ret;
+}
+
+static OSSL_CMP_ITAV *get_genm_itav(OSSL_CMP_CTX *ctx,
+                                    OSSL_CMP_ITAV *req, /* gets consumed */
+                                    int expected, const char *desc)
+{
+    STACK_OF(OSSL_CMP_ITAV) *itavs = NULL;
+    int i, n;
+
+    if (ctx == NULL) {
+        ERR_raise(ERR_LIB_CMP, CMP_R_NULL_ARGUMENT);
+        goto err;
+    }
+    if (OSSL_CMP_CTX_get_status(ctx) != -1
+        /* TODO use OSSL_CMP_PKISTATUS_request when #19205 is merged */) {
+        ERR_raise_data(ERR_LIB_CMP, CMP_R_UNCLEAN_CTX,
+                       "should call CMPclient_reinit() before");
+        goto err;
+    }
+
+    if (!OSSL_CMP_CTX_push0_genm_ITAV(ctx, req))
+        goto err;
+    req = NULL;
+    itavs = OSSL_CMP_exec_GENM_ses(ctx);
+    if (itavs == NULL) {
+        if (1 /*
+               * TODO when PR #19205 is merged: replace '1' by
+               * OSSL_CMP_CTX_get_status(ctx) != OSSL_CMP_PKISTATUS_request
+               */
+            )
+            ERR_raise_data(ERR_LIB_CMP, CMP_R_GETTING_GENP,
+                           "with infoType %s", desc);
+        return NULL;
+    }
+
+    if ((n = sk_OSSL_CMP_ITAV_num(itavs)) <= 0) {
+        ERR_raise_data(ERR_LIB_CMP, CMP_R_INVALID_GENP,
+                       "infoType %s requested, but no ITAV included", desc);
```

Comment:
It looks like `ITAV` is not a standard term in the context of CMP, but rather is OpenSSL-specific jargon.  While it may be reasonable to use in API types, ... where developers will have read the documentation, it should probably not leak into error messages seen by users.

If error messages from this protocol are to be presented to human users, they need a broad overhaul. They are generally only going to be understood by protocol experts, and need to be written for a broader audience.

Reply:
Re-phrased to "response on genm requesting infoType %s does not include suitable value".

> If error messages from this protocol are to be presented to human users, they need a broad overhaul. They are generally only going to be understood by protocol experts, and need to be written for a broader audience.

Most OpenSSL lib error messages are rather cryptic and give pretty little detail such that they are typically hard to make use of, partly even for experts.
When I put error messages like in this PR, I try to be to-the-point and tend to give more detail, 
which hopefully will be help experts, while regular users anyway will not get much out of OpenSSL error queue dumps.

OTOH, the diagnostics of the CMP app should be digestible for regular users,
and with the improvement you suggested above I believe this has been achieved for this PR.